lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
     Value torchCstOne = rewriter.create<Torch::ConstantIntOp>(
         loc, rewriter.getI64IntegerAttr(1));
-    Value expandedIndexTensor = rewriter.create<AtenUnsqueezeOp>(
-        loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    Value torchCstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value indexTensor, flattenedValuesTensorLastDim;
+    BaseTensorType indexTensorType;
+    bool needToFlattenValuesTensor = false;
+    SmallVector<int64_t> flattenedValuesTensorSize;
+    if (indicesList.size() == 1) {
+      indexTensor = indicesList[0];
+      if (indexTensor.getType().isa<Torch::NoneType>())
+        return rewriter.notifyMatchFailure(op,
+                                           "Index tensor must not be None.");
+      indexTensorType = indexTensor.getType().cast<BaseTensorType>();
+      std::optional<unsigned> indexTensorRank = getTensorRank(indexTensor);
+      if (!indexTensorRank)
+        return rewriter.notifyMatchFailure(
+            op, "expected index tensor to have a rank");
+
+      if (*indexTensorRank == 2) {
+        SmallVector<int64_t> indexTensorShape(indexTensorType.getSizes());
+        if (indexTensorShape[0] == 1) {
+          ValueTensorType flattenedIndexTensorType = ValueTensorType::get(
+              context, llvm::ArrayRef({indexTensorShape[1]}),
+              indexTensorType.getDtype());
+          indexTensor = rewriter.create<AtenFlattenUsingIntsOp>(
+              loc, flattenedIndexTensorType, indexTensor, torchCstZero,
+              torchCstOne);
+        } else {
+          return rewriter.notifyMatchFailure(
+              op, "unimplemented: index tensor with rank 2 is supported only "
+                  "when its first dimension is equal to 1");
+        }
+      } else if (*indexTensorRank > 2) {
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: index tensor with rank > 2 is not supported");
+      }
+
+      indexTensorType = indexTensor.getType().cast<BaseTensorType>();
+      assert(indexTensorType.getSizes().size() == 1 &&
+             "expected index tensor to be of rank 1");
+      int64_t indexTensorSize = indexTensorType.getSizes()[0];
+      SmallVector<int64_t> expandedIndexTensorSizes{indexTensorSize, 1};
+      ValueTensorType expandedIndexTensorType = ValueTensorType::get(
+          context, llvm::ArrayRef(expandedIndexTensorSizes),
+          indexTensorType.getDtype());
+      indexTensor = rewriter.create<AtenUnsqueezeOp>(
+          loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    } else if (indicesList.size() > 1 &&
+               (int64_t)indicesList.size() == inputRank) {
+      // Support the cases where index list size is equal to the rank of the
+      // input tensor. Also, the index list should have atmost two non none
+      // index tensors along the consecutive dimension.
+
+      // This implementation assumes input and index tensors to be statically
+      // shaped.
```

Comment:
We should be able to determine the broadcasted shape using dynamic dimensions. We just need to create runtime asserts for dynamic dimensions and create `aten.size.int` ops to get the size of the dynamic dimension.